compiler/frontend/src/mir_optimize/reference_following.rs

```diff
@@ -27,9 +30,10 @@ pub fn follow_references(context: &mut Context, expression: &mut CurrentExpressi
     });
 }
 
-pub fn remove_redundant_return_references(body: &mut Body) {
+pub fn remove_redundant_return_references(body: &mut Body, pureness: &mut PurenessInsights) {
     while let [.., (second_last_id, _), (_, Expression::Reference(referenced))] = &body.expressions[..]
         && referenced == second_last_id {
-        body.expressions.pop();
+        let (id, _) = body.expressions.pop().unwrap();
+        pureness.on_remove(id);
```

Comment:
```suggestion
        pureness.on_remove(id);
        // The expression is a reference, so it can't define any inner IDs we'd
        // have to inform the [`PurenessInsights`] about.
```